gax/src/main/java/com/google/api/gax/retrying/ExponentialRetryAlgorithm.java

```diff
@@ -135,9 +135,16 @@ public boolean shouldRetry(TimedAttemptSettings nextAttemptSettings) {
             - nextAttemptSettings.getFirstAttemptStartTimeNanos()
             + nextAttemptSettings.getRandomizedRetryDelay().toNanos();
 
-    return totalTimeSpentNanos <= globalSettings.getTotalTimeout().toNanos()
-        && (globalSettings.getMaxAttempts() <= 0
-            || nextAttemptSettings.getAttemptCount() < globalSettings.getMaxAttempts());
+    int maxAttempts = globalSettings.getMaxAttempts();
+    long totalTimeout = globalSettings.getTotalTimeout().toNanos();
+
+    // In case maxAttempts is set without setting totalTimeout
+    if (totalTimeout == 0 && maxAttempts > 0) {
+      return nextAttemptSettings.getAttemptCount() < maxAttempts;
+    }
+
+    return totalTimeSpentNanos <= totalTimeout
+        && (maxAttempts <= 0 || nextAttemptSettings.getAttemptCount() < maxAttempts);
```

Comment:
Can we split up the checks for the 2 thresholds? The code is really hard to reason about

something like:

```java
// If totalTimeout limit is defined, check that it hasn't been crossed
if (totalTimeout > 0 && totalTimeSpentNanos > totalTime) {
  return false;
}

// If maxAttempts limit is defined, check that it hasn't been crossed
if (maxAttempts > 0  && nextAttemptSettings.getAttemptCount() >= maxAttempts) {
  return false;
}

// No limits crossed
return true;
```

Reply:
Thanks for the suggestion! Have updated PTAL.